azure/Guidelines.md

```diff
@@ -0,0 +1,864 @@
+# Microsoft Azure REST API Guidelines
+## History
+
+| Date        | Version | Notes                                               |
+| ----------- | ------- | --------------------------------------------------- |
+| 2021-Jul-02 | vNext-RC1 | Updated Azure REST Guidelines.                    |
+| 2020-Jul-31 | v3.2    | Added service advice for initial versions           |
+| 2020-Mar-31 | v3.1    | 1st public release of the Azure REST API Guidelines |
+
+## Introduction
+
+These guidelines offer prescriptive guidance that Azure service teams MUST follow ensuring that customers have a great experience by designing APIs meeting these goals:
+- Developer friendly via consistent patterns & web standards (HTTP, REST, JSON)
+- Efficient & cost-effective
+- Work well with SDKs in many programming languages
+- Customers can create fault-tolerant apps by supporting retries/idempotency/optimistic concurrency
```

Comment:
There is very little documentation in terms of guidance to service owners on how to properly implement idempotency and that has led to unexpected behavior from RP to RP -- e.g. my team had an issue related to an idempotency implementation by a downstream RP where they used a single header value to uniquely identify requests (regardless of content) and return cached responses without doing any further work or returning an error.  Is this guidance coming?  The payment industry has some pretty solid semantics around idempotency that work quite well.  Stripe had a pretty extensive backend around supporting idempotency while also informing customers when idempotency semantics were being violated.

Reply:
@michaelwstark --
We are going to add content and/or a section to the "considerations" document. We'll also look to add a sentence or two here for clarity.

Reply:
I've added a brief discussion of idempotency to the Considerations doc -- starting at line 94 -- which also points into the detailed guidelines in the Guidelines doc.

Reply:
There is ongoing work in the IETF httpapi working group to standardized the Idempotency-Key header used by Stripe. https://github.com/ietf-wg-httpapi/idempotency

Reply:
@michaelwstark -- We've taken the action to reconcile this with the RPs ( see: #273 ). 
@darrelmiller -- we are going to review the Stripe doc and determine if this or the Oasis guidance is what we should include. However, we will do this as a distinct work item (see: #274 )